src/key_packages/mod.rs

```diff
@@ -267,29 +267,65 @@ impl KeyPackageBundle {
     ///
     /// Note that the capabilities extension gets added automatically, based on
     /// the configuration. The ciphersuite for this key package bundle is the
-    /// first one in the `ciphersuites` list. Only the ciphersuites listed in
-    /// `ciphersuites` are enabled in the capabilities extension of the key package.
+    /// first one in the `ciphersuites` list. If a capabilities extension is
+    /// included in the extensions, its supported ciphersuites have to match the
+    /// `ciphersuites` list.
     ///
     /// Returns a new `KeyPackageBundle`.
     pub fn new_with_keypair(
         ciphersuites: &[CiphersuiteName],
         credential_bundle: &CredentialBundle,
-        extensions: Vec<Box<dyn Extension>>,
+        mut extensions: Vec<Box<dyn Extension>>,
         key_pair: HPKEKeyPair,
         leaf_secret: Secret,
     ) -> Result<Self, ConfigError> {
         debug_assert!(!ciphersuites.is_empty());
-        let capabilities_extension = CapabilitiesExtension::new(None, Some(ciphersuites), None);
-        let mut final_extensions: Vec<Box<dyn Extension>> = vec![Box::new(capabilities_extension)];
 
+        // Remove any duplicate extensions
+        extensions.sort();
+        extensions.dedup();
+
+        // Check if the `extensions` already contain the mandatory extensions
+        // and in case one of them is a capabilities extension, add the input
+        // ciphersuites. If not, add the default versions of the mandatory
+        // extensions.
+
+        // First, check if one of the input extensions is a capabilities
+        // extension. If there is, check if one of the extensions is a
+        // capabilities extensions and if the contained ciphersuites are the
+        // same as the ciphersuites passed as input. If that is not the case,
+        // return an error. If none of the extensions is a capabilities
+        // extension, create one that supports the given ciphersuites and that
+        // is otherwise default.
+
+        match extensions
+            .iter()
+            .find(|e| e.get_type() == ExtensionType::Capabilities)
+        {
+            Some(extension) => {
+                let capabilities_extension = extension.to_capabilities_extension().unwrap();
+                if capabilities_extension.ciphersuites() != ciphersuites {
+                    return Err(ConfigError::InvalidCapabilitiesExtension);
+                }
+            }
+
+            None => extensions.push(Box::new(CapabilitiesExtension::new(
+                None,
+                Some(ciphersuites),
+                None,
+            ))),
+        };
+
+        // Check if there is a lifetime extension. If not, add the default one.
+        if !extensions
+            .iter()
+            .any(|e| e.get_type() == ExtensionType::Lifetime)
+        {
+            extensions.push(Box::new(LifetimeExtension::default()));
```

Comment:
`Default` is derived here for `LifetimeExtensions`, so it will be `0` and hence an invalid lifetime.